### DISCUSSION TITLE
how many partitios will zero stage 3  divide mode

### DISCUSSION THREAD

**Discussion by cccc0der** [deepspeedai/DeepSpeed#3438] [Q&A] [answered]
Hi!

I have 640 GPUs on 80 machines to do some pretrain work now,  I have read the [zero tutorials](https://www.deepspeed.ai/tutorials/zero/) but didn't see any config about number of partitions.

My question is, if I start traing on all 80 machines with zero stage 3(without any model parrallel),  will model-parameters be split to 640(_dp 640?_) parts loaded on each GPU?

Waiting for reply, thank you!

**Comment by tjruwase** [comment, marked as answer]
@cccc0der, yes, the vanilla zero3 will split each parameter across the 640 GPUs. However, we are integrating zero3 improvements that split across a subset of DP GPUs. One such called [miCS](https://github.com/microsoft/DeepSpeed/pull/2964) is already available in the 0.9.2 release. A second algorithm will be out soon, so be on the lookout. 

@samadejacobs, FYI.

**Reply by cccc0der** [reply]
thank you!

**Reply by yunll** [reply]
I use miCS config like this
![image](https://user-images.githubusercontent.com/15276878/236410859-c60596ab-d5f0-48d0-97e4-1181fb0bebe9.png)

I get this error
![image](https://user-images.githubusercontent.com/15276878/236411361-48205e5e-c914-4242-9628-6cf71e893245.png)

**Reply by JianqiaoLu** [reply]
do you solve this problem ? I also come across this,,,

**Comment by tjruwase** [comment]
@cccc0der, please open an issue for this.

**Reply by cccc0der** [reply]
done, please check it. https://github.com/microsoft/DeepSpeed/issues/3485

**Comment by samadejacobs** [comment]
@cccc0der, calling MiCS Init in your code should fix reported error:
`with mics.MiCS_Init(data_parallel_group......)`

**Reply by learning-chip** [reply]
Hit the same bug and found here. Yes the error `AttributeError: 'Parameter' object has no attribute 'comm'` can be fixed by using `with deepspeed.zero.MiCS_Init(config_dict_or_path=deepspeed_config):` rather than `with deepspeed.zero.Init():`. The `config_dict_or_path` is required in order to know `mics_shard_size` at init stage.

This behavior is currently not tested in [test_mics_config.py](https://github.com/microsoft/DeepSpeed/blob/v0.9.2/tests/small_model_debugging/test_mics_config.py). Better add it to unit test?

**Reply by JianqiaoLu** [reply]
basically, how can i alter my code to run like mics_init instead of zero init? do you mean I need to change deepspeed internal code?
